service_config.go

```diff
@@ -0,0 +1,104 @@
+/*
+ *
+ * Copyright 2017 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package grpc
+
+import (
+	"encoding/json"
+	"time"
+
+	"google.golang.org/grpc/grpclog"
+)
+
+func parseTimeout(t *string) (*time.Duration, error) {
+	if t == nil {
+		return nil, nil
+	}
+	d, err := time.ParseDuration(*t)
+	return &d, err
+}
+
+type jsonName struct {
+	Service *string `json:"service,omitempty"`
+	Method  *string `json:"method,omitempty"`
+}
+
+func (j jsonName) String() (string, bool) {
+	if j.Service == nil {
+		return "", false
+	}
+	res := "/" + *j.Service + "/"
+	if j.Method != nil {
+		res += *j.Method
+	}
+	return res, true
+}
+
+type jsonMC struct {
+	Name                    *[]jsonName `json:"name,omitempty"`
+	WaitForReady            *bool       `json:"waitForReady,omitempty"`
+	Timeout                 *string     `json:"timeout,omitempty"`
+	MaxRequestMessageBytes  *int        `json:"maxRequestMessageBytes,omitempty"`
+	MaxResponseMessageBytes *int        `json:"maxResponseMessageBytes,omitempty"`
+}
+
+type jsonSC struct {
```

Comment:
This struct is same as the real `ServiceConfig` struct except for the second field. I prefer to not have this temporary struct, and reuse the existing structs.

`json` does a partial parsing if the struct only contains part of the fields, so we could do `Unmarshal` twice, first time for the service/method name, second time for the remaining configs. The second `Unmarshal` can go directly to the existing `MethodConfig` struct.

https://play.golang.org/p/GCSXcP1yQk
```go
const testJSON = `{"name": [ {"service": "foo", "method": "bar"} ], "waitForReady": true}`

type jsonName struct {
	Service *string `json:"service,omitempty"`
	Method  *string `json:"method,omitempty"`
}

type methodConfigName struct {
	Name *[]jsonName
}

type MethodConfig struct {
	WaitForReady *bool
	Timeout      *string
}

func main() {
	buf := []byte(testJSON)

	var mcn methodConfigName
	err := json.Unmarshal(buf, &mcn) // for name only.
	if err != nil {
		panic(err.Error())
	}
	fmt.Printf("%+v\n", mcn)

	var mc MethodConfig
	err = json.Unmarshal(buf, &mc) // for everything other than name.
	if err != nil {
		panic(err.Error())
	}
	fmt.Printf("%+v\n", mc)
}
```

Reply:
filed names are different and go json package cannot handle it.